### PR TITLE
BLOGS-1288 Feeds

### PR DESCRIPTION
**Pull request by mattgreenham**
https://jira.dev.bbc.co.uk/browse/BLOGS-1288

### CODE REVIEW

**Comment by programmes-developers** [comment]
Can one of the admins verify this patch?

File: src/FeedGenerator/FeedGenerator.php [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+        $feed->setTitle($blog->getName() . ' ' . $this->tr('feed'));
+        $feed->setLink($blogUrl);
+        $feed->setFeedLink($feedUrl, $type);
+        $feed->setDescription($blog->getDescription());
+        $feed->setDateModified($lastModified->getTimestamp());
+        $feed->setLanguage($lang);
+        $feed->setEncoding('UTF-8');
+
+        foreach ($posts as $post) {
```
Comment (votemike) [inline src/FeedGenerator/FeedGenerator.php]: Maybe extract the innards of the foreach to a method?

File: src/FeedGenerator/FeedGenerator.php [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+            }
+
+            $feed->addEntry($feedItem);
+        }
+
+        $output = $feed->export($type);
+
+        // Yay a nasty hack so we can display html in atom feeds
+        if ($type == 'atom') {
+            $output = preg_replace('!<content .*?>\s*<xhtml:div .*?>(.*?)</xhtml:div>!is', '<content type="html">$1', $output);
```
Comment (votemike) [inline src/FeedGenerator/FeedGenerator.php]: Can we have a test for this to help people understand WTF.

File: src/FeedGenerator/FeedGenerator.php [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+use App\BlogsService\Domain\Post;
+use App\Ds\PresenterFactory;
+use App\Translate\TranslateProvider;
+use App\Translate\TranslatableTrait;
+use Cake\Chronos\Chronos;
+use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
+use Twig_Environment;
+use Zend\Feed\Writer\Feed;
+
+class FeedGenerator
```
Comment (votemike) [inline src/FeedGenerator/FeedGenerator.php]: I think a test for this file may be useful.

File: src/Ds/Post/ContentBlock/SocialBlock/SocialBlockPresenter.php [hunk 4]
```diff
@@ -25,7 +25,7 @@ public function __construct(Social $content, array $options = [])
 
     public function getAlt(): string
     {
-        return $this->content->getAlt() ?? $this->content->getLink();
+        return $this->content->getAlt() ? $this->content->getAlt() : $this->content->getLink();
```
Comment (votemike) [inline src/Ds/Post/ContentBlock/SocialBlock/SocialBlockPresenter.php]: Tests the break before this change but work afterwards?

File: src/Ds/SidebarModule/UpdatesPresenter.php [hunk 5]
```diff
@@ -30,7 +30,7 @@ public function getBlogId(): string
 
     public function getTagId(): string
     {
-        return $this->tag ? (string) $this->tag->getFileId() : '';
+        return $this->tag ? $this->tag->getId() : '';
```
Comment (votemike) [inline src/Ds/SidebarModule/UpdatesPresenter.php]: Tests the break before this change but work afterwards?

**Comment by programmes-developers** [comment]
Can one of the admins verify this patch?

File: src/Ds/Post/ContentBlock/SocialBlock/SocialBlockPresenter.php (hunk 4 above)
Comment (mattgreenham) [reply src/Ds/Post/ContentBlock/SocialBlock/SocialBlockPresenter.php]: Added

File: src/Ds/SidebarModule/UpdatesPresenter.php (hunk 5 above)
Comment (mattgreenham) [reply src/Ds/SidebarModule/UpdatesPresenter.php]: Added

File: src/FeedGenerator/FeedGenerator.php (hunk 2 above)
Comment (mattgreenham) [reply src/FeedGenerator/FeedGenerator.php]: Added

File: src/FeedGenerator/FeedGenerator.php [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+        if ($post->getAuthor()) {
+            $feedItem->addAuthor(['name' => $post->getAuthor()->getName()]);
+        }
+
+        $postContent = $this->generatePostFeedContent($post);
+
+        if ($type == 'atom') {
+            $postContent = html_entity_decode($postContent, ENT_QUOTES, 'UTF-8');
+            $postContent = htmlspecialchars($postContent, ENT_QUOTES, 'UTF-8');
```
Comment (BjornTwachtmann) [inline src/FeedGenerator/FeedGenerator.php]: Que?

**Pull request merged**